javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll

```diff
@@ -0,0 +1,267 @@
+/**
+ * Provides classes and predicates used by the XSS queries.
+ */
+
+import javascript
+
+/** Provides classes and predicates shared between the XSS queries. */
+module Shared {
+  /** A data flow source for XSS vulnerabilities. */
+  abstract class Source extends DataFlow::Node { }
+
+  /** A data flow sink for XSS vulnerabilities. */
+  abstract class Sink extends DataFlow::Node {
+    /**
+     * Gets the kind of vulnerability to report in the alert message.
+     *
+     * Defaults to `Cross-site scripting`, but may be overriden for sinks
+     * that do not allow script injection, but injection of other undesirable HTML elements.
+     */
+    string getVulnerabilityKind() { result = "Cross-site scripting" }
+  }
+
+  /** A sanitizer for XSS vulnerabilities. */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A regexp replacement involving an HTML meta-character, viewed as a sanitizer for
+   * XSS vulnerabilities.
+   *
+   * The XSS queries do not attempt to reason about correctness or completeness of sanitizers,
+   * so any such replacement stops taint propagation.
+   */
+  class MetacharEscapeSanitizer extends Sanitizer, DataFlow::MethodCallNode {
+    MetacharEscapeSanitizer() {
+      getMethodName() = "replace" and
+      exists(RegExpConstant c |
```

Comment:
This sanitizer does not recognize the string target variant: `tainted.replace('<', ...)` (which is safe in a loop). Perhaps we should make a library for identifying these sanitizing calls to `replace` and friends (other PR), that could be useful for other security queries as well (`".."` for `js/path-injection`, for example).

Reply:
Indeed, but since we're aiming to make the sanitiser relatively permissive anyway I don't think there is an urgent need to cover that case.